File: src/api/hangang.go
package api

import (
	"fmt"

	"github.com/devproje/project-api/src/utils"
)

// Request URL from seoul public api
const HANGANG_URL = "http://openapi.seoul.go.kr:8088/sample/json/WPOSInformationTime"

type Hangang struct {
	WaterPOS struct {
		Row []struct {
			Date   string `json:"MSR_DATE"`
			Time   string `json:"MSR_TIME"`
			SiteID string `json:"SITE_ID"`
			Temp   string `json:"W_TEMP"`
			PH     string `json:"W_PH"`
		} `json:"row"`
	} `json:"WPOSInformationTime"`
}

// Get han river's data
//
//	Area code
//	- 1: Tancheon
//	- 2: Jungnangchecon
//	- 3: Anyangcheon
//	- 4: Seonyudo
//	- 5: Noryangjin
func GetHangang(area int) (*Hangang, error) {
	data, err := utils.GET[Hangang](fmt.Sprintf("%s/%d/%d", HANGANG_URL, area, area))
	if err != nil {
		return nil, err
	}

	for i, d := range data.WaterPOS.Row {
		rawDate := d.Date
		day := fmt.Sprintf("%c%c", rawDate[len(rawDate)-2], rawDate[len(rawDate)-1])
		month := fmt.Sprintf("%c%c", rawDate[len(rawDate)-4], rawDate[len(rawDate)-3])
		year := fmt.Sprintf("%c%c%c%c", rawDate[0], rawDate[1], rawDate[2], rawDate[3])

		data.WaterPOS.Row[i].Date = fmt.Sprintf("%s-%s-%s", year, month, day)
	}

	return data, err
}

File: src/main/kotlin/net/projecttl/api/Hangang.kt
package net.projecttl.api

import io.ktor.utils.io.charsets.*
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import net.projecttl.model.HData
import net.projecttl.model.HangangData
import java.net.URL
import java.nio.charset.Charset

class Hangang {
    private val RAW_URL = "http://openapi.seoul.go.kr:8088/sample/json/WPOSInformationTime"
    fun getData(area: Int = 2): HangangData {
        val start = System.currentTimeMillis()
        val url = URL("$RAW_URL/$area/$area")
        val obj = Json.parseToJsonElement(url.readText(Charset.forName("UTF-8")))
        val data = obj.jsonObject["WPOSInformationTime"]!!.jsonObject["row"]!!.jsonArray[0]

        val resTime = System.currentTimeMillis() - start
        return HangangData(
            status = 200,
            area = data.jsonObject["SITE_ID"].toString().replace("\"", ""),
            data = HData(
                ph = data.jsonObject["W_PH"].toString().replace("\"", ""),
                temp = data.jsonObject["W_TEMP"].toString().replace("\"", ""),
                date = data.jsonObject["MSR_DATE"].toString().replace("\"", ""),
                time = data.jsonObject["MSR_TIME"].toString().replace("\"", "")
            ),
            res_time = "${resTime}ms"
        )
    }
}